nomos-sdp/src/state.rs

```diff
@@ -0,0 +1,615 @@
+        event_type: EventType,
+    ) -> Result<ActiveState, InactiveStateError> {
+        match event_type {
+            EventType::Reward => {
+                self.0.rewarded = Some(block_number);
+                Ok(ActiveState(self.0))
+            }
+            EventType::Declaration | EventType::Withdrawal => {
+                Err(InactiveStateError::ToActiveInvalidEvent(event_type))
+            }
+        }
+    }
+
+    const fn try_into_withdrawn<ContractAddress>(
+        mut self,
+        block_number: BlockNumber,
+        event_type: EventType,
+        service_params: &ServiceParameters<ContractAddress>,
+    ) -> Result<WithdrawnState, InactiveStateError> {
+        match event_type {
+            EventType::Withdrawal => {
+                if self.0.created.wrapping_add(service_params.lock_period) >= block_number {
+                    return Err(InactiveStateError::ToWithdrawalWhileLocked);
+                }
+                self.0.withdrawn = Some(block_number);
+                Ok(WithdrawnState(self.0))
+            }
+            EventType::Declaration | EventType::Reward => {
+                Err(InactiveStateError::ToWithdrawalInvalidEvent(event_type))
+            }
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum WithdrawnStateError {
+    #[error("Withdawn state can be rewarded only during the same blcok")]
+    ToWithdrawnRewardedNotSameBlock,
+    #[error("Withdrawn can not transition to withdrawn rewarded during {0:?} event")]
+    ToWithdrawnRewardedInvalidEvent(EventType),
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct WithdrawnState(ProviderInfo);
+
+impl WithdrawnState {
+    fn try_into_withdrawn_rewarded(
+        mut self,
+        block_number: BlockNumber,
+        event_type: EventType,
+    ) -> Result<Self, WithdrawnStateError> {
+        if matches!(event_type, EventType::Reward) {
+            // Only allow reward for withdrawal in the same block number.
+            if self.0.withdrawn != Some(block_number) {
+                return Err(WithdrawnStateError::ToWithdrawnRewardedNotSameBlock);
+            }
+
+            self.0.rewarded = Some(block_number);
+            Ok(self)
+        } else {
+            Err(WithdrawnStateError::ToWithdrawnRewardedInvalidEvent(
+                event_type,
+            ))
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ProviderStateError {
+    #[error(transparent)]
+    Active(#[from] ActiveStateError),
+    #[error(transparent)]
+    Inactive(#[from] InactiveStateError),
+    #[error(transparent)]
+    Withdrawn(#[from] WithdrawnStateError),
+    #[error("Withdrawn can not transition to active")]
+    WithdrawnToActive,
+    #[error("Provided block is in past, can not transition to state in previous block")]
+    BlockFromPast,
+}
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub enum ProviderState {
+    Active(ActiveState),
+    Inactive(InactiveState),
+    Withdrawn(WithdrawnState),
+}
```

Comment:
Not saying this is the wrong way to do it. But couldn't we encode it using a type state pattern instead? Are we sure all this errors running around are always due to runtime errors? Or are just because we are encoding all kind of transitions?
I would like to understand it a bit better. Maybe a graph with the state machine transition would be helpful in the module documentation.

Reply:
It's totally doable with the typestate pattern, but I fear that we'll endup with a large number of weird states like `ActiveRewarded`, `ActiveNotRewarded`, `InactiveRewarded` etc, so chose partial typestate. Even with full list of states we'll need an enum that combines all of them, because we derive the state from the `ProviderInfo` and we are not sure what state it is during the time the message is received.

Reply:
Ok, lets sync about this. I would like to deeply understand this. Thanks.